src/utils.ts

```diff
@@ -138,66 +152,263 @@ export const updateResourceObject = (original: Resource,
+    //    Object.keys(resource.relationships)
+    //       .forEach(relation => {
+    //           let data = resource.relationships[relation].data;
+    //           if (_.isPlainObject(data)) {
+    //               // hasOne relation
+    //               newState = updateOrInsertResource(state, data, resourceState);
+    //           } else if (_.isArray(data)) {
+    //                // hasMany relation
+    //                newState = <NgrxJsonApiStoreData>data.reduce(
+    //                    (partialState: NgrxJsonApiStoreData, currentResource: Resource): NgrxJsonApiStoreData => {
+    //                        return updateOrInsertResource(partialState, currentResource, resourceState);
+    //                    }, newState);
+    //            }
+    //        });
+    //}
 
     if (_.isUndefined(state[resource.type])) {
         // we must mutate the main state (ngrxjsonapistoredata)
         newState[resource.type] = {};
-        newState[resource.type] = insertResource(newState[resource.type], resource);
+        newState[resource.type] = insertStoreResource(newState[resource.type], resource, fromServer);
         return newState;
 
-    } else if (_.isUndefined(state[resource.type][resource.id])) {
-        newState[resource.type] = insertResource(
-            newState[resource.type],
-            resource);
+    } else if (_.isUndefined(state[resource.type][resource.id]) || override) {
+        newState[resource.type] = insertStoreResource(
+            newState[resource.type], resource, fromServer);
         return newState;
     } else {
-        newState[resource.type] = updateResource(
-            newState[resource.type],
-            resource,
-            state[resource.type][resource.id]);
+        newState[resource.type] = updateStoreResource(
+            newState[resource.type], resource, fromServer);
         return newState;
     }
 
 };
 
+/**
+ * Updates the state of a resource in the store.
+ *
+ * @param state
+ * @param resourceId
+ * @param resourceState
+ * @param loading
+ * @returns {NgrxJsonApiStoreData}
+ */
+export const updateResourceState = (state: NgrxJsonApiStoreData,
+    resourceId: ResourceIdentifier, resourceState? : ResourceState, loading? : boolean): NgrxJsonApiStoreData => {
+    if (_.isUndefined(state[resourceId.type]) || _.isUndefined(state[resourceId.type][resourceId.id])) {
+        return state;
+    }
+    let newState: NgrxJsonApiStoreData = Object.assign({}, state);
+    newState[resourceId.type] = Object.assign({}, newState[resourceId.type]);
+    newState[resourceId.type][resourceId.id] = Object.assign({}, newState[resourceId.type][resourceId.id]);
+    if(resourceState != null){
+        newState[resourceId.type][resourceId.id].state = resourceState;
+    }
+    if(loading != null){
+        newState[resourceId.type][resourceId.id].loading = loading;
+    }
+    return newState;
+};
+
+export const cloneQueryParams = (queryParams: QueryParams): QueryParams => {
```

Comment:
Since we already use lodash, what about `_.cloneDeep`: https://lodash.com/docs/#cloneDeep